kernel: Add k_heap_calloc

sorru94 opened the pull request:
Add the `k_heap_calloc` function. Equivalent to the standard C `calloc` function.

Closes #79417.

cfriedt commented on subsys/tracing/sysview/SYSVIEW_Zephyr.txt:
```diff
@@ -82,6 +82,7 @@ TaskState 0xBF 1=dummy, 2=Waiting, 4=New, 8=Terminated, 16=Suspended, 32=Termina
 
 76  k_heap_init                heap=%I, mem=%p, bytes=%u
 77  k_heap_alloc               heap=%I, bytes=%u, Timeout=%TimeOut | Returns %p
+77  k_heap_calloc              heap=%I, num=%u, size=%u, Timeout=%TimeOut | Returns %p
```
Can this change be added in a separate commit? It's not exactly part of the implementation

cfriedt replied:
Thanks!

dcpleung commented on kernel/kheap.c:
```diff
@@ -116,6 +116,25 @@ void *k_heap_alloc(struct k_heap *heap, size_t bytes, k_timeout_t timeout)
+		ret = k_heap_alloc(heap, bounds, timeout);
+	}
+	if (ret != NULL) {
+		(void)memset(ret, 0, bounds);
```
This may not work correctly if the allocation alignment is bigger than `(num * size)`. For example, you need 2 bytes but the alignment is 8 bytes (pointer size for 64-bit system), this memset will not clear the last 6 bytes.

sorru94 replied:
I see that `k_heap_alloc` is a wrapper for `k_heap_aligned_alloc` with `align` set to  `sizeof(void *)`.
I would use `sizeof(void *)` for memsetting. 
What do you think about it?

npitre commented:
On Tue, 8 Oct 2024, Simone Orru wrote:

> I see that `k_heap_alloc` is a wrapper for `k_heap_aligned_alloc` with `align` set to  `sizeof(void *)`.
> I would use `sizeof(void *)` for memsetting. 
> What do you think about it?

I think this is wrong. In fact I'm not sure I understand the initial 
comment either.

You should zero the requested memory not more. The user is not expected 
toconsume more memory than it asked. The fact that the allocator may 
round up the allocation is an internal detail that no one should rely 
upon.

sorru94 commented:
> On Tue, 8 Oct 2024, Simone Orru wrote: I see that `k_heap_alloc` is a wrapper for `k_heap_aligned_alloc` with `align` set to `sizeof(void *)`. I would use `sizeof(void *)` for memsetting. What do you think about it?
> I think this is wrong. In fact I'm not sure I understand the initial comment either. You should zero the requested memory not more. The user is not expected toconsume more memory than it asked. The fact that the allocator may round up the allocation is an internal detail that no one should rely upon.

I can revert the change. But let's first make sure which one of the two is the correct implementation.

npitre commented:
On Tue, 8 Oct 2024, Daniel Leung wrote:

> Part of me thinks not zeroing out the whole block may have security implications. Maybe it's not an issue.

Security critical code should normally zero its memory before freeing 
it. You can't count on the next memory requestor to do it for you. And 
in that case, too, you're not supposed to access beyond what you 
initially asked for.

calloc() is just a convenience for getting already zeroed memory.

cfriedt reviewed (changes requested):
There seems to be quite a bit of unnecessary code-formatting changes. Please only change code that is relevant to the PR.

npitre commented:
> E.g. the case that Daniel pointed out regarding `memset()`.

As mentioned, that particular case is wrong.

sorru94 commented:
> There seems to be quite a bit of unnecessary code-formatting changes. Please only change code that is relevant to the PR.

The format changes in `subsys/tracing/sysview/tracing_sysview_ids.h` are all related to the alignment of defines. 
Since my changes are only to some of those defines if I only perform alignment for my changes it will result in a header which is partially aligned.

cfriedt commented:
Sure - as long as it's a separate commit, that's ok.

cfriedt reviewed (dismissed):
LGTM

peter-mitsis commented on tests/kernel/mem_heap/k_heap_api/src/test_kheap_api.c:
```diff
@@ -265,3 +267,27 @@ ZTEST(k_heap_api, test_k_heap_alloc_pending_null)
+ *
+ * @ingroup kernel_kheap_api_tests
+ *
+ * @details The test allocates 256 unsigned integers of 32 bytes for a
```
Nit: I think you mean 32 bits (or 4 bytes)

peter-mitsis reviewed (dismissed):
One minor nit, but I don't think that is a stopper.

carlescufi commented:
@sorru94 can you please rebase?

sorru94 committed:
kernel: Add k_heap_calloc

Add the k_heap_calloc function. The
equivalent of the standard calloc for
the Zephyr kernel.

Signed-off-by: Simone Orru <simone.orru@secomind.com>

sorru94 committed:
kernel: Add tests for k_heap_calloc

Add tests for the k_heap_calloc function.

Signed-off-by: Simone Orru <simone.orru@secomind.com>

sorru94 committed:
tracing: Add calloc

Add the calloc tracing functions.

Signed-off-by: Simone Orru <simone.orru@secomind.com>

sorru94 committed:
tracing: sysview: Clang-format tracing_sysview_ids.h

Run clang-format over the tracing_sysview_ids.h.

Signed-off-by: Simone Orru <simone.orru@secomind.com>

kartben commented:
@andyross @peter-mitsis please help review
